rust/src/commands/transaction.rs

```diff
@@ -0,0 +1,171 @@
+    ) -> Self
+    where
+        T: Into<String>,
+    {
+        Self {
+            table_uri: table_uri.into(),
+            input: Arc::new(CoalescePartitionsExec::new(input)),
+            operation,
+            app_metadata,
+        }
+    }
+
+    /// Arrow schema expected to be produced by wrapped operations
+    pub fn input_operation_schema() -> ArrowSchemaRef {
+        Arc::new(OPERATION_SCHEMA.clone())
+    }
+}
+
+#[async_trait]
+impl ExecutionPlan for DeltaTransactionPlan {
+    /// Return a reference to Any that can be used for downcasting
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn schema(&self) -> ArrowSchemaRef {
+        Arc::new(OPERATION_SCHEMA.clone())
+    }
+
+    fn children(&self) -> Vec<Arc<dyn ExecutionPlan>> {
+        vec![self.input.clone()]
+    }
+
+    fn output_partitioning(&self) -> Partitioning {
+        Partitioning::UnknownPartitioning(1)
+    }
+
+    fn required_child_distribution(&self) -> Distribution {
+        Distribution::UnspecifiedDistribution
+    }
+
+    fn with_new_children(
+        &self,
+        _children: Vec<Arc<dyn ExecutionPlan>>,
+    ) -> DataFusionResult<Arc<dyn ExecutionPlan>> {
+        todo!()
+    }
+
+    async fn execute(&self, _partition: usize) -> DataFusionResult<SendableRecordBatchStream> {
+        let mut table =
+            get_table_from_uri_without_update(self.table_uri.clone()).map_err(to_datafusion_err)?;
+
+        let mut actions = Vec::new();
+        let data = collect(self.input.clone()).await?;
+        for batch in data {
+            // TODO we assume that all children send a single column record batch with serialized actions
+            let mut new_actions = deserialize_actions(&batch)?;
+            actions.append(&mut new_actions);
+        }
+
+        if actions.is_empty() {
+            let empty_plan = EmptyExec::new(false, self.schema());
+            return empty_plan.execute(0).await;
+        }
+
+        let _new_version = match table.update().await {
+            Err(_) => {
+                let mut txn = table.create_transaction(None);
+                txn.add_actions(actions);
+                let prepared_commit = txn
+                    .prepare_commit_with_info(
+                        Some(self.operation.clone()),
+                        self.app_metadata.clone(),
+                    )
+                    .await
+                    .map_err(to_datafusion_err)?;
+                let committed_version = table
+                    .try_commit_transaction(&prepared_commit, 0)
+                    .await
+                    .map_err(to_datafusion_err)?;
+                committed_version
+            }
+            _ => {
+                let mut txn = table.create_transaction(None);
+                txn.add_actions(actions);
+                let committed_version = txn
+                    .commit_with_info(Some(self.operation.clone()), self.app_metadata.clone())
```

Comment:
The `commit_with_info` method was actually not meant to live very long. Main reason it exists ritght now, was to propose a way how we do basic commit info using the operation but provide a way for callers to pass in custom application data to be included in the commit. When I saw that `kafka-delta-ingest` uses the commit function I wanted to not break the curretn API for now, but the hope was to change the function signature of the current `commit` method. The current `create` method on delta table also does lower level commit logic. If we decide to use the operations enum as a basis for creating commit infos, create should - I think - use that route as well. And if the call sites in kafka ingest can be adjusted, my haope was to move to a single commit function on the transaction again.

Reply:
Sounds good, we can perform the operation check within `commit_with_info`. My understanding for the current `create` method not needing the operation enum is because there is no commit loop within it, i.e. it only tries to perform a commit for version 0.